main.go

```diff
@@ -12,429 +13,264 @@ import (
+	STATS_WARN := pflag.Float64("stats-warning", 5, "Port stats warning threshold")
+	STATS_CRIT := pflag.Float64("stats-critical", 20, "Port stats critical threshold")
 
+	portsToCheck := pflag.IntSlice("port", []int{1, 2, 3, 4, 5, 6, 7, 8}, "Ports to check")
 	help := pflag.BoolP("help", "h", false, "Show this help")
 
-	// Parse all the flags for later usage
 	pflag.Parse()
 
-	// Display help if the -h is given, or no username / password is provided
 	if *help || *username == "" || *password == "" {
 		pflag.Usage()
 		return
 	}
 
-	*hostName += "/api/v1"
-
-	// Trying to log in
-	err := login(*username, *password)
-	if err != nil { // Error is present, display it
+	n := netgear.NewNetgear(*hostName, *username, *password)
+	if err := n.Login(); err != nil {
 		fmt.Printf("Error while trying to login: %v\n", err)
 		return
 	}
+	defer func() { _ = n.Logout() }()
 
 	if slices.Contains(*mode, "all") {
 		*mode = append(*mode, "basic", "ports", "poe")
 	}
 
-	// worstStatus is needed for 'inheriting' the worst status from the lower levels to the top
 	worstStatus := check.OK
 
-	// Now everything is fine and the token is saved in a global variable in functions.go
-	// We can proceed with checking for every mode checked and doing corresponding stuff
-
-	//Basic output
+	// Basic check
 	if slices.Contains(*mode, "basic") {
-		var data map[string]any
-		inputData := device_info()
-		err = json.Unmarshal(inputData, &data)
+		var deviceInfo netgear.DeviceInfo
+		inputData, err := n.DeviceInfo()
 		if err != nil {
-			panic(err)
+			fmt.Printf("Error retrieving device info: %v\n", err)
+			os.Exit(int(check.Unknown))
+		}
+		if err := json.Unmarshal(inputData, &deviceInfo); err != nil {
+			fmt.Printf("Failed to parse JSON: %v\n", err)
+			os.Exit(int(check.Unknown))
 		}
-		deviceInfo := data["deviceInfo"].(map[string]any)
-
-		// Basic info
-		upTime := deviceInfo["details"].([]any)[0].(map[string]any)["upTime"]
-		cpuUsage := string_percent_to_float(deviceInfo["cpu"].([]any)[0].(map[string]any)["usage"].(string))
-		memoryUsage := string_percent_to_float(deviceInfo["memory"].([]any)[0].(map[string]any)["usage"].(string))
-
-		// Fans details
-		fanDetails := deviceInfo["fan"].([]any)[0].(map[string]any)["details"].([]any)[0].(map[string]any)
-		fanName := fanDetails["desc"].(string)
-		fanSpeed := fanDetails["speed"].(float64)
 
-		// Temperature details
-		sensorDetails := deviceInfo["sensor"].([]any)[0].(map[string]any)["details"].([]any)
+		upTime := deviceInfo.DeviceInfo.Details[0].Uptime
+		cpuUsage := netgear.StringPercentToFloat(deviceInfo.DeviceInfo.Cpu[0].Usage)
+		memUsage := netgear.StringPercentToFloat(deviceInfo.DeviceInfo.Memory[0].Usage)
+		fan := deviceInfo.DeviceInfo.Fan[0].Details[0]
+		fanName := fan.Description
+		fanSpeed := fan.Speed
+		sensorDetails := deviceInfo.DeviceInfo.Sensor[0].Details
 
-		// Create result container
 		o := result.Overall{}
 
-		// CPU check
+		// CPU
 		if !*hidecpu {
```

Comment:
All these checks - CPU, Memory, Temperature, and so on - could be extracted into their own function, which is then just called here. This would help making the `main` function more readable. Furthermore, this might show repeated patterns for a further refactoring.